### PR TITLE
WIP - Switch to golanci-lint

### PR DESCRIPTION
**Pull request by TomSweeneyRedHat**
Signed-off-by: TomSweeneyRedHat <tsweeney@redhat.com>

Attempt to switch to golangci-lint which is replacing
gometalinter, which is getting deprecated soon.

Stealing heavily from https://github.com/kubernetes-sigs/cri-o/pull/2137
by @saschagrunert

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -70,15 +74,14 @@ endif
 	@./tests/validate/whitespace.sh
 	@./tests/validate/govet.sh
 	@./tests/validate/git-validation.sh
-	@./tests/validate/gometalinter.sh . cmd/buildah
+	lint
```
Comment (TomSweeneyRedHat) [inline Makefile]: I think I'll probably need to make an equivalent gometalinter.sh for the new lint.  Will work on this more tomorrow.

**Comment by rhatdan** [comment]
@TomSweeneyRedHat Any movement on this?

**Comment by TomSweeneyRedHat** [comment]
@vrothberg FYI, if I get out from under BZ's while you're gone, I'll try to dive back into this.

**Comment by rhatdan** [comment]
Perhaps @saschagrunert Would like to help with this.  HintHint...

**Comment by vrothberg** [comment]
> Perhaps @saschagrunert Would like to help with this. HintHint...

Feel free to cherry-pick from https://github.com/containers/buildah/pull/1624.

File: .golanc.yml [hunk 2]
```diff
@@ -0,0 +1,38 @@
+---
```
Comment (saschagrunert) [inline .golanc.yml]: The file has the wrong name and should be `.golangci.yml`

File: Makefile [hunk 3]
```diff
@@ -50,6 +50,10 @@ clean:
 docs: ## build the docs on the host
 	$(MAKE) -C docs
 
+.PHONY: lint
+lint: ## do lint processing 
+	golangci-lint run --build-tags="$(BUILDTAGS)"
```
Comment (saschagrunert) [inline Makefile]: I would hard-code the buildtags to produce the same output on every developer machine, like here: 
https://github.com/cri-o/cri-o/blob/97c3b029ad69d7347520e6efc932ac27c8659c33/.golangci.yml#L3-L8

File: Makefile (hunk 1 above)
Comment (saschagrunert) [reply Makefile]: I guess a single invocation of `golangci-lint run` could be enough here.

**Comment by saschagrunert** [comment]
If we rename the file, then I only see some issues left on my machine:
```
> make lint
golangci-lint run --build-tags="seccomp selinux    containers_image_ostree_stub"
bind/util.go:29:6: `dedupeStringSlice` is unused (deadcode)
func dedupeStringSlice(slice []string) []string {
     ^
chroot/run_test.go:107:6: `testNoop` is unused (deadcode)
func testNoop(t *testing.T) {
     ^
chroot/run_test.go:114:6: `testMinimalSkeleton` is unused (deadcode)
func testMinimalSkeleton(t *testing.T) {
     ^
chroot/util.go:5:6: `dedupeStringSlice` is unused (deadcode)
func dedupeStringSlice(slice []string) []string {
     ^
make: *** [Makefile:55: lint] Error 1
```

We could also exclude `deadcode` for now and enable the linters later on including the fixes.

**Comment by rhatdan** [comment]
SGTM

File: Makefile (hunk 3 above)
Comment (TomSweeneyRedHat) [reply Makefile]: I'd like to get @nalind's opinion on this.  He set the calls to a variety of shell scripts to get the value for the platform.  I'm not sure if we can hardcode it without messing up a particular variant on linux.

Comment (saschagrunert) [reply Makefile]: That's an argument, so I guess we can also stick to the current approach.

**Comment by TomSweeneyRedHat** [comment]
Replaced by #1686

**Pull request closed without merging**